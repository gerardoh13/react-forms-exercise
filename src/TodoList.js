import React, { useState } from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

const TodoList = () => {
  const [todos, setTodos] = useState([]);
  const addTodo = (newTodo) => {
    setTodos((prev) => [...prev, newTodo]);
  };

  const deleteTodo = (id) => {
    setTodos((prev) => prev.filter((t) => t.id !== id));
  };

  const toggleDone = (id) => {
    let todosCopy = [...todos]
    const found = todosCopy.find(t => t.id === id);
    found.done = !found.done
    setTodos(todosCopy)
  }
  
  const editTask = (id, task) => {
    let todosCopy = [...todos]
    const found = todosCopy.find(t => t.id === id);
    found.todo = task
    setTodos(todosCopy)
  }

  const todoList = todos.map((t, idx) => (
    <Todo
      key={t.id}
      id={t.id}
      index={idx}
      todo={t.todo}
      done={t.done}
      deleteTodo={deleteTodo}
      toggleDone={toggleDone}
      editTask={editTask}
    />
  ));
  return (
    <>
      <NewTodoForm addTodo={addTodo} />
      {todos.length === 0 ? null : (
        <div className="notePad mt-3"> {todoList}</div>
      )}
    </>
  );
};

export default TodoList;
