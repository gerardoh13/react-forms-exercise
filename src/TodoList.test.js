import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

function addTodo(component, val) {
  const taskInput = component.getByLabelText("Todo:");
  fireEvent.change(taskInput, { target: { value: val } });
  const submitBtn = component.getByText("Add!");
  fireEvent.click(submitBtn);
}

it("renders without crashing", function () {
  render(<TodoList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a todo", function () {
  const todoList = render(<TodoList />);
  addTodo(todoList, "TEST TODO");
  expect(todoList.getByText("TEST TODO")).toBeInTheDocument();
  expect(todoList.getByText("✎")).toBeInTheDocument();
  expect(todoList.getByText("✔")).toBeInTheDocument();
  expect(todoList.getByText("x")).toBeInTheDocument();
});

it("can edit a todo", function () {
  const todoList = render(<TodoList />);
  addTodo(todoList, "TEST TODO");
  const editBtn = todoList.getByText("✎");
  fireEvent.click(editBtn);
  const editIpt = todoList.queryByTestId("editIpt");
  fireEvent.change(editIpt, { target: { value: "NEW TEST VALUE" } });
  const saveBtn = todoList.getByText("Save");
  fireEvent.click(saveBtn);
  expect(todoList.getByText("NEW TEST VALUE")).toBeInTheDocument();
});

it("can delete a todo", function () {
    const todoList = render(<TodoList />);
    addTodo(todoList, "TEST TODO");
    const todo = todoList.getByText("TEST TODO")
    expect(todo).toBeInTheDocument();
    const delBtn = todoList.getByText("x");
    fireEvent.click(delBtn);
    expect(todo).not.toBeInTheDocument();
  });