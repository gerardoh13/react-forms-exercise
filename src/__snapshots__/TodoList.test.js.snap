// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class=" row todoForm"
  >
    <div
      class="col-11"
    >
      <div
        class="form-floating"
      >
        <input
          class="form-control"
          id="todo"
          name="todo"
          placeholder="todo"
          required=""
          type="text"
          value=""
        />
        <label
          class="label"
          for="todo"
        >
          Todo:
        </label>
      </div>
    </div>
    <div
      class="col-1"
    >
      <button
        class="btn btn-warning btn-lg"
      >
        Add!
      </button>
    </div>
  </form>
</DocumentFragment>
`;
